### PR TITLE
provide SPI spring namespace configuration for sharding algorithms

### LINKED ISSUE #5196: Provide SPI spring namespace configuration for all the Sharding algorithms
Hi 
**Content:** This issue is to provide SPI spring namespace configuration for Sharding algorithms. 
**Difficulty rating:** ★★☆☆☆

### PR DESCRIPTION
**Pull request by strongduanmu**
Fixes #5196 .

Changes proposed in this pull request:
- provide SPI spring namespace configuration for sharding algorithms

### CODE REVIEW

File: sharding-spring/sharding-jdbc-orchestration-spring/sharding-jdbc-orchestration-spring-namespace/src/test/resources/META-INF/rdb/namespace/shardingMasterSlaveNamespace.xml [hunk 1]
```diff
@@ -32,12 +32,12 @@
...
-    
-    <bean id="standardModuloDatabaseShardingAlgorithm" class="org.apache.shardingsphere.shardingjdbc.orchestration.spring.algorithm.StandardModuloDatabaseShardingAlgorithm" />
-    <bean id="standardModuloTableShardingAlgorithm" class="org.apache.shardingsphere.shardingjdbc.orchestration.spring.algorithm.StandardModuloTableShardingAlgorithm" />
-    <bean id="defaultComplexKeysShardingAlgorithm" class="org.apache.shardingsphere.shardingjdbc.orchestration.spring.algorithm.DefaultComplexKeysShardingAlgorithm" />
-    <bean id="defaultHintShardingAlgorithm" class="org.apache.shardingsphere.shardingjdbc.orchestration.spring.algorithm.DefaultHintShardingAlgorithm" />
-   
+
+    <spi:sharding-algorithm id="standardModuloDatabaseShardingAlgorithm" type="STANDARD_TEST_DB"/>
+    <spi:sharding-algorithm id="standardModuloTableShardingAlgorithm" type="STANDARD_TEST_TBL"/>
+    <spi:sharding-algorithm id="defaultComplexKeysShardingAlgorithm" type="complex"/>
```
Comment (tristaZero) [inline sharding-spring/sharding-jdbc-orchestration-spring/sharding-jdbc-orchestration-spring-namespace/src/test/resources/META-INF/rdb/namespace/shardingMasterSlaveNamespace.xml:38]: `complex` is not a proper type name, so could rename it in `defaultComplexKeysShardingAlgorithm ` as `COMPLEX_TEST`.

**Comment by tristaZero** [comment]
@strongduanmu Hi I like your PR!
It impressed me much that you can finish this task neatly in the short term. BTW, could you look at my comment to make this PR perfect?

File: sharding-spring/sharding-jdbc-orchestration-spring/sharding-jdbc-orchestration-spring-namespace/src/test/resources/META-INF/rdb/namespace/shardingMasterSlaveNamespace.xml (hunk 1 above)
Comment (strongduanmu) [reply sharding-spring/sharding-jdbc-orchestration-spring/sharding-jdbc-orchestration-spring-namespace/src/test/resources/META-INF/rdb/namespace/shardingMasterSlaveNamespace.xml:38]: > `complex` is not a proper type name, so could rename it in `defaultComplexKeysShardingAlgorithm ` as `COMPLEX_TEST`.

@tristaZero Okay, I will modify it.

**Comment by strongduanmu** [comment]
> @strongduanmu Hi I like your PR!
> It impressed me much that you can finish this task neatly in the short term. BTW, could you look at my comment to make this PR perfect?

@tristaZero Hi, I have modified it, please review again, thank you! 😀

**Comment by strongduanmu** [comment]
> @strongduanmu Hi I like your PR!
> It impressed me much that you can finish this task neatly in the short term. BTW, could you look at my comment to make this PR perfect?

@tristaZero Hi, I need some help. After I commit the modified code, the ci task got an error. It seems to be a problem with coveralls-maven-plugin, but I don't know how to fix it.

image

**Comment by tristaZero** [comment]
Hi I triggered CI to run twice more, but the results are same,
```
[INFO] BUILD FAILURE
35736[INFO] ------------------------------------------------------------------------
35737[INFO] Total time:  19:15 min
35738[INFO] Finished at: 2020-04-18T06:09:01Z
35739[INFO] ------------------------------------------------------------------------
35740[ERROR] Failed to execute goal org.eluder.coveralls:coveralls-maven-plugin:4.3.0:report (default-cli) on project shardingsphere: Processing of input or output data failed: Report submission to Coveralls API failed with HTTP status 422: Unprocessable Entity (Couldn't find a repository matching this job.) -> [Help 1]
35741[ERROR] 
35742[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
35743[ERROR] Re-run Maven using the -X switch to enable full debug logging.
35744[ERROR] 
35745[ERROR] For more information about the errors and possible solutions, please read the following articles:
35746[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
35747/home/travis/.travis/functions: line 607:  4565 Terminated              travis_jigger "${!}" "${timeout}" "${cmd[@]}"
35748The command "travis_wait 30 mvn --batch-mode --no-transfer-progress clean install cobertura:cobertura coveralls:report -DrepoToken="${COVERALLS_REPO_TOKEN}" -Dmaven.javadoc.skip=true" exited with 1.
cache.2
35749store build cache
35763
afte
```
@strongduanmu Have you ever installed successfully in you local environment?

**Comment by strongduanmu** [comment]
> Hi I triggered CI to run twice more, but the results were same,
> 
> ```
> [INFO] BUILD FAILURE
> 35736[INFO] ------------------------------------------------------------------------
> 35737[INFO] Total time:  19:15 min
> 35738[INFO] Finished at: 2020-04-18T06:09:01Z
> 35739[INFO] ------------------------------------------------------------------------
> 35740[ERROR] Failed to execute goal org.eluder.coveralls:coveralls-maven-plugin:4.3.0:report (default-cli) on project shardingsphere: Processing of input or output data failed: Report submission to Coveralls API failed with HTTP status 422: Unprocessable Entity (Couldn't find a repository matching this job.) -> [Help 1]
> 35741[ERROR] 
> 35742[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
> 35743[ERROR] Re-run Maven using the -X switch to enable full debug logging.
> 35744[ERROR] 
> 35745[ERROR] For more information about the errors and possible solutions, please read the following articles:
> 35746[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
> 35747/home/travis/.travis/functions: line 607:  4565 Terminated              travis_jigger "${!}" "${timeout}" "${cmd[@]}"
> 35748The command "travis_wait 30 mvn --batch-mode --no-transfer-progress clean install cobertura:cobertura coveralls:report -DrepoToken="${COVERALLS_REPO_TOKEN}" -Dmaven.javadoc.skip=true" exited with 1.
> cache.2
> 35749store build cache
> 35763
> afte
> ```
> 
> @strongduanmu Have you ever installed with the lastest commit successfully in your local environment?

@tristaZero In my local environment, the latest commit code can successfully execute `mvn install`. I noticed that the pr merged today has a similar error log. The following is the corresponding ci error log. Is there any other ci configuration problem?
* https://travis-ci.org/github/apache/shardingsphere/builds/676499292?utm_source=github_status&utm_medium=notification
* https://travis-ci.org/github/apache/shardingsphere/builds/676481674?utm_source=github_status&utm_medium=notification

**Comment by tristaZero** [comment]
@strongduanmu Got it. Sorry for that. I will reflect this ISSUE to my colleagues, please wait a moment.

**Comment by tristaZero** [comment]
@strongduanmu 
Hi, It seems there is something wrong with the CI test. But given your successful case, I merged it.
Thanks for your important contribution to the community. Please consider interacting more, as you are needed here. :-)

**Comment by strongduanmu** [comment]
> @strongduanmu
> Hi, It seems there is something wrong with the CI test. But given your successful case, I merged it.
> Thanks for your important contribution to the community. Please consider interacting more, as you are needed here. :-)
Hi @tristaZero, I am willing to participate more in community work, because the tolerance and enthusiasm of the SS community deeply attracted me. BTW, I want to participate in the issue #4683 , can you assign me a subtask?

**Comment by tristaZero** [comment]
@strongduanmu I am appreciated your attitude for SS community. BTW, is #4683 your only option, or are there any other subjects you have interest in?

**Comment by terrymanu** [comment]
@strongduanmu Thank you for your compliments, welcome and hope we can do more things together.

**Comment by strongduanmu** [comment]
> @strongduanmu I am appreciated your attitude for SS community. BTW, is #4683 your only option, or are there any other subjects you have interest in?

Hi, @tristaZero , I am currently learning the SS parser module and trying to understand ANTLR, so I hope to claim related tasks. I am also paying attention to the tasks of other modules. If it is within my ability, I will participate in it actively.

**Comment by strongduanmu** [comment]
> @strongduanmu Thank you for your compliments, welcome and hope we can do more things together.

@terrymanu I will try to participate more in community tasks.😀

**Pull request merged**